statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,292 @@
+        Returns
+        -------
+        est : VARResults
+        """  # TODO: docstring + implementation
+        
+        # TODO: trend
+        
+        # select number of lags (=p-1)
+        if ic is None:
+            if max_diff_lags is None:
+                diff_lags = 1
+            else:
+                diff_lags = max_diff_lags
+        else:
+            selections = self.select_order(max_diff_lags=max_diff_lags,
+                                           verbose=verbose)
+            if ic not in selections:
+                raise ValueError("%s not recognized, must be among %s"
+                                 % (ic, sorted(selections)))
+            diff_lags = selections[ic]
+            if verbose:
+                print("Using %d based on %s criterion" %  (diff_lags, ic))
+        
+        # estimate parameters
+        if method == "ls":
+            return self._estimate_vecm_ls(diff_lags, deterministic)
+        elif method == "egls":
+            if coint_rank is None:
+                coint_rank = 1
+            return self._estimate_vecm_egls(diff_lags, deterministic,
+                                            coint_rank)
+        elif method == "ml":
+            if coint_rank is None:
+                coint_rank = 1
+            return self._estimate_vecm_ml(diff_lags, deterministic,
+                                          coint_rank)
+        else:
+            raise ValueError("%s not recognized, must be among %s"
+                             % (method, ("ls", "egls", "ml")))
+
+    def _est_matrices(self, diff_lags, deterministic):
+        # p. 286:
+        p = diff_lags+1
+        y = self.endog.T  # superclass turning DataFrame into ndarray?
+        K = y.shape[0]
+        y_1_T = y[:, p:]
+        T = y_1_T.shape[1]
+        delta_y = np.diff(y)
+        delta_y_1_T = delta_y[:, p-1:]
+        y_min1 = y[:, p-1:-1]
+        if "lt" in deterministic:
+            y_min1 = vstack((y_min1,
+                             np.arange(T)))  # p. 299
+        # p. 286:
+        delta_x = np.zeros((diff_lags*K, T))
+        for j in range(delta_x.shape[1]):
+            delta_x[:, j] = (delta_y[:, j+p-2:None if j-1<0 else j-1:-1]
+                             .T.reshape(K*(p-1)))
+        # p. 299, p. 303:
+        if "c" in deterministic:
+            delta_x = vstack((delta_x,
+                              np.ones(T)))
+        if "s" in deterministic:  # TODO: How many seasons??
+            seasons = np.zeros((3, delta_x.shape[1]))  # 3 = 4-1 (4=#seasons)
+            for i in range(seasons.shape[0]):
+                seasons[i, i::4] = 1
+            #seasons = seasons[:, ::-1]
+            #seasons = np.hstack((seasons[:, 3:4], seasons[:, :-1]))
+            #seasons = np.hstack((seasons[:, 2:4], seasons[:, :-2]))
+            seasons = np.hstack((seasons[:, 1:4], seasons[:, :-3]))
+            # seasons[1] = -seasons[1]
+            delta_x = vstack((delta_x,
+                              seasons))
```

Comment:
the `if` above need to be refactored, I think constant/trends should not be separately specified, i.e. use `elif` for different constant and trend specifications (Stata has 5).
repeated `vstack` will make several copies of the data, It's better to keep them in a list and then use one np.column_stack